lumicks/pylake/kymo.py

```diff
@@ -261,39 +261,65 @@ def plot_with_time_histogram(self, color_channel, pixels_per_bin=1, hist_ratio=0
+    def downsampled_by(
+        self, time_factor=1, position_factor=1, reduce=np.sum, reduce_timestamps=np.mean
+    ):
         """Return a copy of this Kymograph which is downsampled by `time_factor`
 
         Parameters
         ----------
         time_factor : int
-            The number of pixels that will be averaged in time.
+            The number of pixels that will be averaged in time (default: 1).
+        position_factor : int
+            The number of pixels that will be averaged in space (default: 1).
         reduce : callable
             The `numpy` function which is going to reduce multiple pixels into one.
             The default is `np.sum`.
+        reduce_timestamps : callable
+            The `numpy` function that is going to reduce timestamps into one.
+            The default is `np.mean`.
         """
         result = Kymo(self.name, self.file, self.start, self.stop, self._json)
 
         def image_factory(_, channel):
             data = self._image(channel)
-            return block_reduce(data, (1, time_factor), func=reduce)[
-                :, : data.shape[1] // time_factor
+            return block_reduce(data, (position_factor, time_factor), func=reduce)[
+                : data.shape[0] // position_factor, : data.shape[1] // time_factor
             ]
 
-        def timestamp_factory(_):
+        def timestamp_factory_ill_defined(_):
             raise AttributeError(
-                "Per-pixel timestamps are no longer available after downsampling a kymograph since "
-                "they are not well defined (the downsampling occurs over a non contiguous time "
-                "window). Line timestamps are still available however. See: "
+                "Per-pixel timestamps are no longer available after downsampling a kymograph in "
+                "time since they are not well defined (the downsampling occurs over a "
+                "non-contiguous time window). Line timestamps are still available, however. See: "
                 "`Kymo.line_time_seconds`."
             )
 
+        def timestamp_factory(_):
+            return block_reduce(self.timestamps, (position_factor, 1), func=reduce_timestamps)[
+                : self.timestamps.shape[0] // position_factor, :
+            ]
+
         def line_time_factory(_):
             return self.line_time_seconds * time_factor
 
+        def pixelsize_factory(_):
+            pixelsizes = self.pixelsize_um
+            pixelsizes[0] = pixelsizes[0] * position_factor
+            return pixelsizes
+
+        def pixelcount_factory(_):
+            num_pixels = self._num_pixels
+            num_pixels[0] = num_pixels[0] // position_factor
+            return num_pixels
+
         result._image_factory = image_factory
-        result._timestamp_factory = timestamp_factory
+        result._timestamp_factory = (
+            timestamp_factory_ill_defined if time_factor != 1 else timestamp_factory
```

Comment:
I'd write it exactly the other way around, to have the happy path for `time_factor == 1` and the ill defined default for the else branch.

Reply:
I had it this way because `time_factor` downsampling would be the most frequently used path. But I'm fine with turning the logic around. :)